### ISSUE TITLE
Push notifications do not work (SQL syntax error)

### ISSUE THREAD

**Issue by errotu** [mattermost/mattermost#29459]
#### Summary
Push notifications do not work on Android and iOS, probably due to a SQL syntax error.

#### Steps to reproduce
1. User A sends a message to user B.
2. User B expects push notification on the other end.
3. User B does not receive a push notification.

#### Operating System
CentOS7

#### Mattermost Server Version
10.2.0

#### Expected behavior
Push notifications should be sent to the mobile devices.

#### Observed behavior
Mobile devices do not receive push notifications (iOS and Android alike). 

#### Log Output
Hundreds of entries like this:

> {"timestamp":"2024-11-28 20:53:55.332 +01:00","level":"error","msg":"Failed to send mobile app sessions","caller":"app/notification_push.go:123","logSource":"notifications","type":"push","status":"error","reason":"fetch_error","user_id":"3875fqztif8exe4oo8c1enwd8o","error":"getMobileAppSessions: We encountered an error while finding user sessions., failed to find Sessions with userId=3875fqztif8exe4oo8c1enwd8o: Error 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '>>'$.last_removed_device_id', '')' at line 9"}

#### Additional Information
The same behavior (with the same error messages) occurs on two different Mattermost instances I maintain. They are completely independent, but are both using the same web hosting service ([uberspace](https://lab.uberspace.de/guide_mattermost/)). Push notifications worked until a couple of weeks ago.

**Comment by agnivade** [comment]
@errotu/ @r1de - We don't support MariaDB. This is a syntax issue with MariaDB.

**Comment by errotu** [comment]
I thought that MySQL/MariaDB (as far as I know, MariaDB is completely compatible with the MySQL syntax) [are supported until Mattermost v11](https://docs.mattermost.com/upgrade/important-upgrade-notes.html)?

> We no longer support new installations using MySQL starting in v10. All new customers and/or deployments will only be supported with the minimum supported version of the PostgreSQL database. End of support for MySQL is targeted for Mattermost v11.

**Comment by errotu** [comment]
Just fyi: For the moment, I downgraded our server back to 9.11.5 and push notifications work again. However, if v10 does not support MySQL/MariaDB, this should be indicated clearly.

**Comment by agnivade** [comment]
We support MySQL. Not MariaDB. MariaDB is not compatible with MySQL. As you can see clearly in the error.

**Comment by r1de** [comment]
It is true that MariaDB and MySQL are 100% compatible only 99% of the time. :smile:  

That explains it, but the error does not show that clearly.  

After reading this, I successfully migrated my MM instance to Postgres today, thereby solving this notification issue on my end (and anything else that might pop up later due to MariaDB).  
Ran into a few issues during the migration process, which took several hours and a few failed attempts to figure it all out, but after working all that out it was successful.

**Comment by ruhnet** [comment]
@r1de mind sharing how you did the migration, and what hurdles you encountered? Might help some of us in the same boat avoid the same issues. :)

**Comment by errotu** [comment]
> We support MySQL. Not MariaDB. MariaDB is not compatible with MySQL. As you can see clearly in the error.

You're right, @agnivade, the "completely compatibility" of both systems was an overstatement. However, it seems like at least until v10.2 compatibility was given and Mattermost worked with both systems equally. 

The error only indicates that the current version does not work flawlessly with MariaDB, it does not necessarily mean that it works with MySQL (and, thus, is just a problem of compatibility). Are you certain the error does not occur using MySQL?

If I'm not mistaken, the issue was introduced by #28269 and most probably [along these lines](https://github.com/mattermost/mattermost/pull/28269/files#diff-9aa7cdc79bfaaab8dc77ac51328692d91f196debec55b451fe36895af493b395R149) (based on the error message and the time the problem occurred first):

```
	lastRemovedQuery := `DeviceId != COALESCE(Props->>'last_removed_device_id', '')`
	if me.DriverName() == model.DatabaseDriverMysql {
		lastRemovedQuery = `DeviceId != COALESCE(Props->>'$.last_removed_device_id', '')`
	}
```

I'm not skilled enough to identify the exact problem, but more generally I did not find any reports about an incompatibility between [MySQL COALESCE](https://www.w3schools.com/Sql/func_mysql_coalesce.asp) and [MariaDB COALESCE](https://mariadb.com/kb/en/coalesce/), so I'm a bit skeptical that this is really a compatibility issue.

**Comment by agnivade** [comment]
This is not a problem with COALESCE. This is a problem with JSON interoperability in MySQL and MariaDB. They are not 100% compatible.

**Comment by r1de** [comment]
@ruhnet I started here: https://docs.mattermost.com/deploy/postgres-migration-assist-tool.html
You will also need to setup the postgres DB initially, use this page: https://docs.mattermost.com/install/prepare-mattermost-database.html

Follow those.   
The postgres DB has to be setup before you start running the migration tools on it.
If you do not have any playbooks or boards, you can skip those parts in Section 6 of the migration page.
You will have to do the "calls" part of Section 6 still though, do not forget that part.

Once you setup your MM to use the new Postgres DB, you may get this error in the Mattermost Server log:  
```
sql: Scan error on column index 0, name \"current_schema\": converting NULL to string is unsupported, query: \n\nSELECT CURRENT_SCHEMA()\n"
```
To fix that, first check a few things:
Login to postgres _as_ the Mattermost user: `psql -U mmuser -d mattermostdb`
Run: `select current_schema();` it'll likely be blank
Run `show search_path;` and if you see something like `"""""%user%""", public""""` then log out of postgres and log back in as the super user (ie: `sudo -u postgres psql`) and run the following command (changed for your site of course)
`ALTER USER mmuser IN DATABASE mattermostdb SET search_path TO 'public';`


Hopefully that'll help somewhat and I am remembering correctly.
I was going fast when I was doing the migration and figuring it out, so I probably do not remember _everything_ I did to get it working.
This page also helped: https://github.com/mattermost/migration-assist/issues/16
Just remember that your are _not_ destroying the MySQL/Maria DB, your just copying it, so if something goes wrong, you can go back and still use it.
Also remember to just drop the postgres DB and start over if you have to.

**Comment by addyess** [comment]
@larkox 

i found by changing the query from 
```
SELECT * FROM Sessions WHERE UserId="REDACTED" AND ExpiresAt != 0 AND DeviceId != '' AND DeviceId != COALESCE(Props->>'$.last_removed_device_id', '');
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '>>'$.last_removed_device_id', '')' at line 1
```

to 

```sql
SELECT * FROM Sessions WHERE UserId="REDACTED" AND ExpiresAt != 0 AND DeviceId != '' AND DeviceId != COALESCE(JSON_UNQUOTE(JSON_EXTRACT(Props, '$.last_removed_device_id')), '');
 ```
 
 the query does work on mariadb -- not sure if this is compatible with MySQL
 
 Perhaps this is a more compatible query?

**Comment by agnivade** [comment]
Thank you @addyess. That latter does work with MySQL. But unfortunately, we do not support MariaDB. Making one-off fixes to support both DBs gives off the impression that we do, and becomes an endless exercise to keep fixing MariaDB bugs whenever they keep coming up.

More importantly, we are close to dropping MySQL support as well. So all of this is entirely going to go away soon, allowing us to focus on just one DB - Postgres.

I'd highly suggest to migrate to Postgres whenever you can. I am going to close the issue now as it's been open for quite some time. Thanks all.

**Comment by addyess** [comment]
Wow, i just migrated to mysql from mariadb.  It wasn't difficult to dump from one then load into another.  Using postgresql seems a bit heavy handed.  When is this mysql deprecation coming?

**Comment by errotu** [comment]
Thanks a lot for providing a fix, @addyess!

Would it be possible to reconsider if this fix couldn't be applied to Mattermost v10, @agnivade? While you're right that Mattermost does not officially support MariaDB, it _de facto_ worked with MariaDB until this change was introduced. Given that MySQL won't be supported from v11 on, there's also only very little danger to establish a precedent for further issues.

In the spirit of open source, it would be really great if this community fix for a community use (also considering it's only a single line of code) could be applied. It seems like a minor change for the Mattermost project, but gives MariaDB users (which are probably in many cases smaller, voluntary-driven projects) some more time for the Postgres migration (if I understand the [Release Policy](https://docs.mattermost.com/about/release-policy.html#extended-support-releases) correct, v10.11will be supported until May 2026).

If I'm not mistaken, the fix could be applied by changing line 151 of server/channels/store/sqlstore/session_store.go [like here](https://github.com/mattermost/mattermost/commit/1368c42bb26404e60c9b3eae0953de6e7acaa65c?diff=split&w=0).

**Comment by errotu** [comment]
@addyess, according to the [v10 changelog](https://docs.mattermost.com/about/mattermost-v10-changelog.html#id26) MySQL support will be dropped completely in v11:

> We no longer support new installations using MySQL starting in v10. All new customers and/or deployments will only be supported with the minimum supported version of the PostgreSQL database. End of support for MySQL is targeted for Mattermost v11.

**Comment by addyess** [comment]
@errotu aye aye.  I've got a bit of time i suppose.  I'm running a mattermost instances on a raspberry pi 4 sharing double duty with pihole.  Something feels odd about needing postgres for something 8 users connect to. But I get it -- "production" is such a fun word to say that we spend alot of effort as developers getting it right.  And I understand wanting to optimize DB queries to a specific SQL instance type.  Postgres is the right one in this case.

**Comment by agnivade** [comment]
> While you're right that Mattermost does not officially support MariaDB, it de facto worked with MariaDB until this change was introduced. 

That is not really correct. It might have been the case for you, but we've had a lot of different issues with customers running MariaDB before. We did accept small fixes in the past, but that just led to the impression that we support MariaDB, leading us to fix even more issues. And since we never really test with MariaDB in CI, it's more work for us to keep fixing these one-off edge cases with MariaDB.

I completely understand that making this small one-line fix lets you upgrade to newer versions at very low cost, but we have been bitten by this in the past by trying to keep accepting such small one-liner fixes. One solution for you might be to simply move to MySQL for now. Migrating from MariaDB to MySQL is relatively simpler, and should allow you to use newer versions until you have the time to move to Postgres.

Hopefully, this is not a major inconvenience for you. If you run into any issues while migrating to Postgres, please don't hesitate to reach out to us. We will be glad to help you out.

**Comment by errotu** [comment]
Thank you for all the background information! I was finally able to take some time for the migration to PostgreSQL. It was a bit more tedious than I had hoped, but in the end it all worked out!

One general suggestion: Could you link the [Documentation about the environment configuration](https://docs.mattermost.com/configure/environment-configuration-settings.html#database) at Step 7 of the [migration how-to](https://docs.mattermost.com/deploy/postgres-migration-assist-tool.html)? Without that information it's not really clear how the SqlSettings.DataSource and SqlSettings.DriverName values have to be modified.

Also, [when preparing the PostgreSQL database](https://docs.mattermost.com/install/prepare-mattermost-database.html), I had to add the following command (as 6.d) to make it work:
> ALTER SCHEMA public OWNER TO mmuser;

## Some more specific information related to my set-up
Those might be interesting for administrators in a similar situation but are probably not relevant for the general documentation:

- Our set-up is that we run Mattermost on an [uberspace](https://lab.uberspace.de/guide_mattermost/) and used a separate Hetzner Cloud (Ubuntu 24.04) _(only)_ for the migration of the database

**PostgreSQL version**
- As uberspace [only supports PostgreSQL up to v15](https://lab.uberspace.de/guide_postgresql/), it is important [to install PostgreSQL v15](https://linuxcapable.com/install-postgresql-15-on-ubuntu/) (and not a later version) on the Hetzner Server to be able to later import the migrated PostgreSQL database on uberspace

**pgloader**
- I was not able to install the mattermost-pgloader on the Hetzner-Server ("docker: no matching manifest for linux/arm64/v8 in the manifest list entries.")
- the regular pgloader from the repositories didn't work at a certain stage but threw me an "76 fell through ECASE expression" error
- compiling pgloader by myself [as described here](https://stackoverflow.com/a/76648875) worked fine
- in consequence, I had to change Step 4 of the migration to:
> pgloader/build/bin/pgloader migration.load > migration.log

**Comment by agnivade** [comment]
Thanks @errotu ! Would you be open to sending a PR? The page is at https://github.com/mattermost/docs/blob/master/source/deploy/postgres-migration-assist-tool.rst.

**Comment by atverdokhlib** [comment]
HI @addyess sorry for question where I can change sql query? Where can find file on on-premise server. 
Thank you.

**Comment by addyess** [comment]
@atverdokhlib it was in the [go code](https://github.com/mattermost/mattermost/pull/28269/files#diff-9aa7cdc79bfaaab8dc77ac51328692d91f196debec55b451fe36895af493b395R149) and require a rebuild of the code to implement.  Not a config change you can make on-prem.

**Comment by atverdokhlib** [comment]
@addyess if I migrate to PostgreSQL  issue will be fixed ?

**Comment by addyess** [comment]
@atverdokhlib , yes i migrated from maridb to mysql.  I had to do a db migration from one to the other which went well for my very small installation. postgres while more complicated would also work i believe.  

sorry, i don't actually maintain this project, but just happened to see a way around the contributors decided to not pickup

**Comment by agnivade** [comment]
@atverdokhlib - If you migrate to Postgres, it will be resolved. Migration guide here: https://docs.mattermost.com/deploy/postgres-migration.html
